### PR TITLE
[Nano] Support users to choose `jit.trace`/`jit.script` in `InferenceOptimizer.trace`/`quantize`

### PR DESCRIPTION
**Pull request by Oscilloscope98**
## Description

Support users to choose `jit.trace`/`jit.script` when applying `jit` as the accelerator in `InferenceOptimizer.trace`/`quantize`

### 1. Why the change?

When converting a model to TorchScript, users could either use `jit.trace` or `jit.script`, or even mix both of them. These two converting methods have their own prons and cons. For example, for a model that are both scriptable and traceable, `jit.trace` could always generate a simpler graph (thus probably better performance). However, `jit.trace` would cause problems when the model includes dynamic workflows, it could also capture variables as constants, etc. (Reference: this [blog](https://ppwwyyxx.com/blog/2022/TorchScript-Tracing-vs-Scripting/#Let-Tracing-Generalize))

Currently, in `InferenceOptimizer.trace`/`quantize`, we are only using `jit.trace`. For better flexibility, we should enable users to choose either of them. A try-except logic will also be used (first `jit.trace`, failed then `jit.script`) by default to increase the success rate of converting the model.

### 2. User API changes

Added a `jit_method` parameter to  `InferenceOptimizer.trace`/`quantize` for users to choose whether to use `jit.trace` or `jit.srcipt`. By default,  a try-except logic will be used (first `jit.trace`, failed then `jit.script`) .

```python
# use jit.trace
accmodel = InferenceOptimizer.trace(model, accelerator='jit', 
                                    input_sample=input_sample, 
                                    jit_method='trace')

# use jit.script
accmodel = InferenceOptimizer.trace(model, accelerator='jit', 
                                    input_sample=input_sample, 
                                    jit_method='script')

# use try-except logic
accmodel = InferenceOptimizer.trace(model, accelerator='jit', 
                                    input_sample=input_sample)
```

### 3. Summary of the change 
- Added a `jit_method` parameter to  `InferenceOptimizer.trace`/`quantize`
- Added the missing `jit_strict` parameter when loading a `PytorchIPEXJITBF16Model` model

### 4. How to test?
- [x] Unit test
- [x] Document test for related API: https://yuwentestdocs.readthedocs.io/en/nano-jit-script-trace/doc/PythonAPI/Nano/pytorch.html#bigdl.nano.pytorch.InferenceOptimizer.quantize

### 5. New dependencies

N/A

### 6. Possible further improvements
Try to detect/inform the case that `jit.trace` causes problems in `InferenceOptimizer.optimize` due to problems in dynamic workflows, capturing variables as constants, etc,

### CODE REVIEW

File: python/nano/test/pytorch/tests/inference/test_bf16_ipex.py [hunk 1]
```diff
@@ -150,6 +151,65 @@ def hello():
         with pytest.raises(AttributeError):
             new_model.width
 
+    def test_bf16_ipex_jit_converter(self):
```
Comment (rnwang04) [inline python/nano/test/pytorch/tests/inference/test_bf16_ipex.py:154]: You don't have to put two identical unit tests in two different files.
Now `test_bf.py` only contains ut for `bf16` and `bf16+channels_last`, and all `ipex / jit` related ut are put in `test_bf16_ipex.py` and `test_ipex_jit_inference.py`. So just put ut in the last two files.

File: python/nano/src/bigdl/nano/deps/ipex/ipex_api.py [hunk 2]
```diff
@@ -48,16 +48,19 @@ def PytorchIPEXJITModel(model, input_sample=None, use_ipex=False,
     :param thread_num: the thread num allocated for this model.
     :param inplace: whether to perform inplace optimization. Default: ``False``.
     :param jit_strict: Whether recording your mutable container types.
+    :param jit_converter: use ``jit.trace`` or ``jit.script`` to
```
Comment (rnwang04) [inline python/nano/src/bigdl/nano/deps/ipex/ipex_api.py:51]: In my feeling, `jit_converter` is not intuitive and a little strange.

I think `jit_compiler` may be a choice：
_Scripting a function or nn.Module will inspect the source code, compile it as TorchScript code using the TorchScript compiler, and return a [ScriptModule]_\
_Trace a function and return an executable or [ScriptFunction] that will be optimized using just-in-time compilation._

**Review by rnwang04 — COMMENTED** [review]
others LGTM.

File: python/nano/test/pytorch/tests/inference/test_bf16_ipex.py (hunk 1 above)
Comment (Oscilloscope98) [reply python/nano/test/pytorch/tests/inference/test_bf16_ipex.py:154]: Thank you for the reminding! I am kind of confused by the file names. Thank you! Have updated,

File: python/nano/src/bigdl/nano/deps/ipex/ipex_api.py (hunk 2 above)
Comment (Oscilloscope98) [reply python/nano/src/bigdl/nano/deps/ipex/ipex_api.py:51]: I also think that `jit_converter` could be not so intuitive.

However, after some investigation, I think one of the biggest difference between `jit.script` and `jit.trace` is that `jit.script` will "compiles" a model into graph, but `jit.trace` will not.

From the direct explanation in TorchScript official documentation:
image
From my understanding, it also shows that the complication of model converted from `jit.trace` will be conducted **jit in time** (but the complication will not be done by `jit.trace` itself).

So, for me, I feel like that `jit_compiler` will by itself already refer to `jit.script`, so may not be a suitable parameter name.

@TheaperDeng Do you have any idea what will be a more suitable name?

Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/deps/ipex/ipex_api.py:51]: maybe we could just call it `jit_method` after some discussion with @MeouSker77 @rnwang04

Comment (Oscilloscope98) [reply python/nano/src/bigdl/nano/deps/ipex/ipex_api.py]: Have renamed the parameter.

**Review by TheaperDeng — APPROVED** [review]
LGTM

**Pull request merged**